File: contest/leetcode/path-sum-iii.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:

        from collections import Counter
        ans = 0

        def visit(root):
            nonlocal ans
            c = Counter()
            if root is None:
                return Counter()

            lc = visit(root.left)
            rc = visit(root.right)
            v = root.val
            c = Counter()
            for x, c0 in lc.items():
                c[x + v] += c0
            for x, c0 in rc.items():
                c[x + v] += c0
            c[v] += 1

            ans += c[sum]
            return c

        visit(root)
        return ans

File: leetcode/cpp/path-sum-iii.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <cstdio>
struct TreeNode {
    int val;
    TreeNode *left, *right;
};
class Solution {
   public:
    int trace(TreeNode* root, int exp) {
        if (root == NULL) {
            return 0;
        }
        int count = 0;
        if (root->val == exp) {
            count += 1;
        }
        count += trace(root->left, exp - root->val);
        count += trace(root->right, exp - root->val);
        return count;
    }
    int pathSum(TreeNode* root, int sum) {
        if (root == NULL) return 0;
        int count = 0;
        count += trace(root, sum);
        count += pathSum(root->left, sum);
        count += pathSum(root->right, sum);
        return count;
    }
};
